### PR TITLE
Add MetricsSystem.release() to allow metric lifetime management #6

### LINKED ISSUE #6: Consider a form of "unregister"/"destroy" for metrics
While reading this proposal keep the following situation in mind:

- the metrics API, being this library
- a metrics implementation, being an implementation that bootstraps the API and has some metrics handlers, and potentially a "registry" as well.
  - many metrics systems need this kind of registry; e.g. when a poll based one is implemented the `describe()` call has to render all the metrics the application has, to be able to do this, it needs to keep them around in a registry (e.g. prometheus or similar ones)
- some other library, that wants to offer its users metrics; it must not and can not know about the metrics library implementation

In such setup a "some other library" wants to do its best to do the right thing, but it can't know if a metrics library has a concept of lifecycle, if we don't have it in the API here. Even if most of the time it would be noop.

----

Example scenario: 

- Some other library controls lifecycle of things, and it also emits metrics for them. 
- It only knows about this metrics API. 
- The used metrics library allocates heavy histograms and other things for each created metric.
  - It would want to free those once we know we'll never use them again; after they've been send to some gathering system.
- The other library can't call into this "release" without the shared metrics API having it in its api.

Example of metrics clients which would need this:

In prometheus metrics can unregister https://github.com/prometheus/client_golang/blob/master/prometheus/registry.go#L184 they can never be registered again with a different type, but that's fine for the use cases I am considering. Since the library always would use the same types of metrics for the same keys.

I'll PR an example which may clarify things a bit.

I would also as part of this say that we likely do _not_ want to ship the `CachingMetricsHandler` as metrics implementations should implement those on their types in order to make releasing() possible / effective.

### PR DESCRIPTION
**Pull request by ktoso**
Please see https://github.com/tomerd/swift-server-metrics-api-proposal/issues/6 for detailed rationale. In many metrics systems it is good to release metrics resources that are known to be no longer used.

For many metrics this may not matter if they are 'for the application' like "requests per second" etc since those never "go away", however in order to offer more fine grained metrics the number of metrics potentially stored in a metrics library "registry" may grow indefinitely potentially causing what could be seen as a leak. For this purpose a release() mechanism has to be in place in the shared API, in order to allow "middle ware" (any driver, web framework etc), to emit and release metrics whenever they see fit.

Since prometheus is mentioned in metrics discussions in SSWG quite a bit, please note that the release semantics do exist and are useful in Prometheus. So if we aim for that to be one of our backends, that by itself another reason to offer this API -- regardless of the heavy histograms case I point out in the README.

This can be discussed along the way as the API is proposed to become part of the SSWG projects I think @tomerd. Hope we'll nail it down quickly :-)

Resolves https://github.com/tomerd/swift-server-metrics-api-proposal/issues/6
Replaces #7

### CODE REVIEW

File: Sources/CoreMetrics/Metrics.swift [hunk 1]
```diff
@@ -12,12 +12,15 @@
 //
 //===----------------------------------------------------------------------===//
 
+public protocol MetricHandler: AnyObject {
```
Comment (ktoso) [inline Sources/CoreMetrics/Metrics.swift:15]: This unlocks being able to do an API like: `release(metric: MetricHandler)`
Implementations need to `switch` anyway internally since they most likely want to get their internal types ("TestCounterHandler" etc) to get the identity that should be released

File: Sources/CoreMetrics/Metrics.swift [hunk 2]
```diff
@@ -12,12 +12,15 @@
...
+}
+
 /// This is the Counter protocol a metrics library implements
-public protocol CounterHandler: AnyObject {
+public protocol CounterHandler: MetricHandler {
     func increment<DataType: BinaryInteger>(_ value: DataType)
 }
 
-// This is the user facing Counter API. It must have reference semantics, and its behviour depend ons the `CounterHandler` implementation
+// This is the user facing Counter API. It must have reference semantics, and its behavior depend ons the `CounterHandler` implementation
```
Comment (ktoso) [inline Sources/CoreMetrics/Metrics.swift:23]: Used the american spelling to fit rest of style

File: Sources/CoreMetrics/Metrics.swift [hunk 3]
```diff
@@ -165,6 +168,39 @@ public protocol MetricsFactory {
...
+    /// This function MAY be invoked concurrently, and implementations should take care to use appropriate
+    /// synchronization mechanisms where necessary.
+    func release<M: MetricHandler>(metric: M)
+}
+
+extension MetricsFactory {
+    public func release<M: MetricHandler>(metric: M) {
+        // no-op by default.
+        // Libraries which do maintain metric lifecycle should implement this method.
+    }
```
Comment (ktoso) [inline Sources/CoreMetrics/Metrics.swift:203]: So implementations which don't care never have to even implement the no-op.

File: Sources/Examples/ExampleMetricsLibrary.swift [hunk 4]
```diff
@@ -67,6 +67,40 @@ class ExampleMetricsLibrary: MetricsFactory {
...
+            if let idx = self.counters.firstIndex(where: { m in m.label == counter.label }) {
+                lock.withLock {
+                    self.counters.remove(at: idx)
+                }
+            }
+        case let gauge as ExampleGauge:
+            if let idx = self.gauges.firstIndex(where: { m in m.label == gauge.label }) {
+                lock.withLock {
+                    self.gauges.remove(at: idx)
+                }
```
Comment (ktoso) [inline Sources/Examples/ExampleMetricsLibrary.swift:84]: TODO: Hm... should move the lock around the find as well I guess.

File: Sources/Examples/SimpleMetricsLibrary.swift [hunk 5]
```diff
@@ -98,7 +98,7 @@ class SimpleMetricsLibrary: MetricsFactory {
         }
 
         func record<DataType: BinaryFloatingPoint>(_ value: DataType) {
-            // this may loose percision but good enough as an example
+            // this may loose precision but good enough as an example
```
Comment (ktoso) [inline Sources/Examples/SimpleMetricsLibrary.swift:101]: was a typo

File: README.md [hunk 6]
```diff
@@ -206,6 +207,39 @@ class SimpleMetricsLibrary: MetricsFactory {
...
+
+Not all libraries will have a need for this lifecycle management, e.g. if `MetricHandler` implementations are stateless
+forwarders to some backend system, it is likely that managing lifecycle of individual metrics objects may not be required.
+However, in case of libraries which keep any created `MetricHandler` in some internal registry (e.g. in the form of a 
+dictionary `[MetricId: MetricHandler]`), it may be necessary to invoke `MetricsSystem.release()` for known not-used-anymore 
+metrics in order to facilitate eagerly discarding this state from the registry.
+
+The eager releasing of metrics plays an even more important role for `MetricsHandlers` which rely on "heavy" resources
+such as: file handles, network connections or large in-memory data structures (e.g. high definition resolution histograms,
+which can take tens, up to hundreds of KiB).
```
Comment (ktoso) [inline README.md:224]: For reference https://github.com/HdrHistogram/HdrHistogram#footprint-estimation and that's a fairly efficient one (without losing precision) 

// C port https://github.com/HdrHistogram/HdrHistogram_c

File: Sources/CoreMetrics/Metrics.swift (hunk 3 above)
Comment (MrLotU) [reply Sources/CoreMetrics/Metrics.swift:203]: I think it might be better to go for explicit denial of lifetime usage, so to say, so that creators of backends are forced to consider if they need to implement this or not

File: Sources/Examples/ExampleMetricsLibrary.swift [hunk 7]
```diff
@@ -99,12 +133,21 @@ class ExampleMetricsLibrary: MetricsFactory {
     }
 }
 
-class ExampleCounter: CounterHandler, CustomStringConvertible {
+/// In this `ExampleMetricsLibrary` we decided to have a shared type for our metrics as we will identify them by the label alone.
+/// Other libraries my implement identification differently; This example only goes to show how libraries can use manage identity (in the simplest case).
```
Comment (MrLotU) [inline Sources/Examples/ExampleMetricsLibrary.swift:137]: Typo
```suggestion
/// Other libraries may implement identification differently; This example only goes to show how libraries can use manage identity (in the simplest case).
```

File: Tests/MetricsTests/CoreMetricsTests.swift [hunk 8]
```diff
@@ -229,4 +229,36 @@ class MetricsTests: XCTestCase {
...
+
+        let gaugeAgain  = Gauge(label: name)
+        gaugeAgain.record(-value)
+
+        let recorderAgain = Gauge(label: name).handler as! TestRecorder
+        XCTAssertEqual(recorderAgain.values.count, 1, "expected number of entries to match")
+        XCTAssertEqual(recorderAgain.values.first!.1, -value, "expected value to match")
+
+        let identityAgain = ObjectIdentifier(recorderAgain)
+        XCTAssertNotEqual(identity, identityAgain, "since the cached metric was released, the created a new should have a different identity")
```
Comment (MrLotU) [inline Tests/MetricsTests/CoreMetricsTests.swift:262]: Typo
```suggestion
        XCTAssertNotEqual(identity, identityAgain, "since the cached metric was released, the newly created one should have a different identity")
```

File: Sources/CoreMetrics/Metrics.swift (hunk 3 above)
Comment (ktoso) [reply Sources/CoreMetrics/Metrics.swift:203]: Sorry for the radio silence here @MrLotU, back at it now 🙇‍♂️

Either way is likely fine... I think Tom's main concern was to not cause adoption friction for super simple implementations, thus this provided no-op. Maybe others on the SSWG have an opinion here too?

**Comment by tomerd** [comment]
@ktoso i made some of the typo / language fixes in https://github.com/tomerd/swift-server-metrics-api-proposal/pull/13

for the lifecycle part, do you think we can make the release function take a label instead of an object? could simplify the API?

File: Sources/CoreMetrics/Metrics.swift (hunk 3 above)
Comment (tomerd) [reply Sources/CoreMetrics/Metrics.swift:203]: /cc @seabaylea

**Comment by ktoso** [comment]
Let me prototype it and PR so we can compare 👍

**Comment by ktoso** [comment]
I gave the `release(label:)` style and here's a commit with the changes to it from the style in this PR (`release(metric:)`): https://github.com/ktoso/swift-server-metrics-api-proposal/commit/7f6e6c53281c1030c6145c02532352f4cabcaaec

I'm not convinced it would simplify things for either users of implementors. More in depth info in: https://github.com/ktoso/swift-server-metrics-api-proposal/commit/7f6e6c53281c1030c6145c02532352f4cabcaaec#r33046627

TL;DR; I think is: 

- If I'm an user or library which cares about managing lifecycle, I do keep references to the metrics objects anyway
  - keeping names is error prone, as they're stringly typed and one has to watch out to pass the "right string" etc, e.g. if a library happens to apply a namespace or something to labels that I passed in.
  - releasing with `M: Metric` still seems to make sense to me, as whoever wants to release things, should have been keeping references to them anyway.
- for users and libraries which do not care about lifecycle, e.g. because the metrics they expose "never become invalid", like global request counters for a process etc; then this API does not impact them.
- trivial metric library implementations which do not contain registries do not need to concern themselfes about this proposal at all.
   - but yes, middle wares do, though this loops back to point 1)

**Comment by ktoso** [comment]
Will rebase, now that https://github.com/tomerd/swift-server-metrics-api-proposal/pull/13#event-2255985230 landed (but tomorrow :-))

**Comment by ktoso** [comment]
This PR will move over to the repositories new location that is being prepared, now that is has been accepted by the SSWG. ( https://forums.swift.org/t/april-4th-2019/22704 )

🕐

**Comment by tomerd** [comment]
now tracked in https://github.com/apple/swift-metrics/pull/3
